core/src/blockchain/blockchain.rs

```diff
@@ -149,6 +170,12 @@ impl BlockChain {
         if let Some(hash) = pending_best_block_hash.take() {
             *best_block_hash = hash;
         }
+
+        let mut highest_block_hash = self.highest_block_hash.write();
+        let mut pending_highest_block_hash = self.pending_highest_block_hash.write();
```

Comment:
I think all the locks should be created at once if all of them should be updated.
How about moving these two lines above a bit?